service/topology_coordinator.cc

```diff
@@ -2478,18 +2478,20 @@ future<bool> topology_coordinator::maybe_start_tablet_split_finalization(group0_
 
 future<locator::load_stats> topology_coordinator::refresh_tablet_load_stats() {
     auto tm = get_token_metadata_ptr();
-    auto& topology = tm->get_topology();
 
     locator::load_stats stats;
     static constexpr std::chrono::seconds wait_for_live_nodes_timeout{30};
 
     std::unordered_map<table_id, size_t> total_replicas;
 
-    for (auto& [dc, nodes] : topology.get_datacenter_nodes()) {
+    for (auto& [dc, endpoints] : tm->get_datacenter_token_owners()) {
         locator::load_stats dc_stats;
-        rtlogger.info("raft topology: Refreshing table load stats for DC {} that has {} endpoints", dc, nodes.size());
-        co_await coroutine::parallel_for_each(nodes, [&] (const auto& node) -> future<> {
-            auto dst = node->host_id();
+        rtlogger.info("raft topology: Refreshing table load stats for DC {} that has {} endpoints", dc, endpoints.size());
+        co_await coroutine::parallel_for_each(endpoints, [&] (const auto& endpoint) -> future<> {
+            auto dst = tm->get_host_id_if_known(endpoint);
```

Comment:
Is it even possible for this to return endpoint with unknown host ID? Maybe we should establish invariants in these APIs. It sounds like if `get_datacenter_token_owners` returns a "token owner" with unknown host ID -- that's kind of a serious problem.

Unless this is supposed to handle some kind of race where the node leaves while we are iterating, then maybe it makes sense.

Reply:
It is impossible. I'll change `get_host_id_if_known` to `get_host_id`.

 `token_metadata_impl::get_datacenter_token_owners` returns only nodes with host IDs. I'll make it clearer by refactoring it a bit. There is this weird `is_normal_token_owner(get_host_id_if_known(ip).value_or(host_id{}))` that could call `is_normal_token_owner` with `host_id{}`.